AdvancedC#/FileSystemVisitorPL/FileSystemVisitor.cs

```diff
@@ -0,0 +1,59 @@
+﻿namespace FileSystemVisitorPL
+{
+    public class FileSystemVisitor : IFileSystemVisitor
+    {
+        private readonly string rootFolder;
+        private readonly Func<string, bool> filter;
+
+        public FileSystemVisitor(string rootPath)
+        {
+            if (!Directory.Exists(rootPath))
+            {
+                throw new DirectoryNotFoundException("The specified root folder does not exist.");
+            }
```

Comment:
It's a good practice to put some common validation rules into separate functions. In this case, you can declare `ValidateRootDirectory` method with `void` type